### PR TITLE
Use combobox for region search

### PR DESCRIPTION
**Pull request by Luceium**
Introduces a combobox with cached location results.
Integrates with Nav.
On selection, sets region.
Allows region search.

### CODE REVIEW

File: components/search/RegionSearch/RegionSelect.tsx
Comment (ray-sjsu) [inline components/search/RegionSearch/RegionSelect.tsx:1]: Hmm... I think you start with the basics first, like the `navbar`. The navbar component will be available for all pages and within the navbar, you will store that information through searchParams.

In `lib/rapid-hotel-api/constants/USER_OPTIONS.ts`, there are the default GLOBAL options for booking/reservations.
Start with the navbar. The navbar would `useSearchParams()` to retrieve any searchParams in the url (if they exist). If they don't exist, fallback to the default options provided in `USER_OPTIONS`.

Now, which variables/information to store in searchParams (if user selected one)? I recommend that these searchParams don't show up in the URL (on first visit) if they are the default option from `USER_OPTIONS`.

```
`region_search` // this is the searchQuery for searching hotels in a particular region.
`region_id`
`domain`
`locale`
`adults_number`

`accessibility`
`amenities`
`meal_plan`
`lodging`
`sort_order`
`min_price`
`max_price`

// these dates require a little bit more work
`checkin_date`
`checkout_date`
```

Some variables are omitted from `USER_OPTIONS.ts`, as we may use the default options instead and will not be changed by the user. The custom API uses `hotelSearchParamsBasicSchema` from `/lib/rapid-hotel-api/zod` to verify valid inputs.

`checkin_date` and `checkout_date` is more tricky. You would have to generate them using a function. Storing static dates on a constants file does allow for dynamic default dates. I generated dates on a range provided by `DEFAULT_BOOKING_NUM_DAYS` and `DEFAULT_DAYS_FROM_TODAY` in `USER_OPTIONS.ts` See the homepage example:
```
  const { DEFAULT_CHECKIN_BOOKING_DATE, DEFAULT_CHECKOUT_BOOKING_DATE } =
    generateDefaultDates(DEFAULT_BOOKING_NUM_DAYS);

  const [searchParams, setSearchParams] = useState<searchParamsType>({
    // RegionSearch default inputs
    query: DEFAULT_QUERY,
    domain: DEFAULT_DOMAIN,
    locale: DEFAULT_LOCALE,
    selectedRegionId: "",

    // BookingInfo default inputs
    checkinDate: DEFAULT_CHECKIN_BOOKING_DATE,
    checkoutDate: DEFAULT_CHECKOUT_BOOKING_DATE,
    adultsNumber: DEFAULT_ADULTS_NUMBER,
    numDays: DEFAULT_BOOKING_NUM_DAYS,

    // HotelSearch default inputs
    accessibilityOptions: DEFAULT_ACCESSIBILITY_OPTIONS,
    amenitiesOptions: DEFAULT_AMENITIES_OPTIONS,
    mealPlanOptions: DEFAULT_MEAL_PLAN_OPTIONS,
    lodgingOptions: DEFAULT_LODGING_OPTIONS,
    paymentType: DEFAULT_PAYMENT_TYPE_OPTIONS,
    sortOrder: DEFAULT_SORT_ORDER,
    availableOnly: DEFAULT_AVAILABILITY_FILTER_OPTIONS,
    price_min: DEFAULT_MIN_PRICE,
    price_max: DEFAULT_MAX_PRICE,
  });
 ```

TLDR: Start with the basics, how are you going to persistently store information between pages? Start with navbar. Since the navbar will work across all pages, use `searchParams` (in the URL) to store the booking information. Hotels.com does something similar with searchParams.

Default values are found in `USER_OPTIONS.ts` file, then move the combobox into the navbar once navbar can actually update and retrieve information from searchParams.

Comment (Luceium) [reply components/search/RegionSearch/RegionSelect.tsx:1]: The majority of your comment is out of scope. This PR aims to replace the simple query with a combo box that can autocomplete cached region IDs.

Comment (ray-sjsu) [reply components/search/RegionSearch/RegionSelect.tsx:1]: Hmm, you're right.

Although, I hope you consider this when creating the Navbar.

I'll review it again.

File: components/search/RegionSearch/RegionSelect.tsx [hunk 1]
```diff
@@ -78,6 +83,7 @@ const RegionSelect: React.FC<RegionSelectProps> = ({
 
   return (
     <div>
+      <LocationCombobox searchLocations={handleFindRegion} setRegionId={onRegionSelect} />
```
Comment (ray-sjsu) [inline components/search/RegionSearch/RegionSelect.tsx:86]:
```suggestion
```

Since this is `Region Search`. We only need `query`, `domain`, `locale`. All of this is available + more in `HomeSearchPage` aka `app/page.tsx`. Remove this from `RegionSelect.tsx` and put it in `app/page.tsx`.

It would be better to make your own API call to `api/hotels/region` within `LocationCombobox`.

In the future, put your custom components in `app/page.tsx` to have full access to all booking info.

File: components/ui/location-combobox.tsx
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:1]: Here, make your own API call to `/api/hotel/region`. An example exists within `RegionSelect.tsx`.
- There is a function `handleFindRegion` which takes `query`, `domain`, and `locale` as parameters and retrieves data in the format of `APIRegionArrayFormatted`.
- There is a useState hook called `regionSearchOutput` that stores a list of regions (initially `null`).
- There is a useState hook called `selectedRegionDetails`. This is for one region only and is selected by the user or selected by default (if you want). It is selected from the array provided by `regionSearchOutput`.

Use the default values in `/lib/constants/USER_OPTIONS.ts` when initializing `LocationCombobox` for the first time. From what I can tell, the search bar is not pre-populated with "San Jose" which is the value of `DEFAULT_QUERY` in `USER_OPTIONS.ts`. `DEFAULT_DOMAIN` for domain and `DEFAULT_LOCALE` for locale if you want to implement those (they are optional for the API call).

Refer to the original code. TypeScript is quite strict with its typing.

Ultimately, the end goal of `LocationCombobox` should be to retrieve a valid `region_id` that can be passed onto other components and for further API calls (like getting a hotel from a region).

**Review by ray-sjsu — COMMENTED** [review]
Re-review

File: components/search/RegionSearch/RegionSelect.tsx (hunk 1 above)
Comment (Luceium) [reply components/search/RegionSearch/RegionSelect.tsx:86]: That's bad practice. Components should be abstracted and organized into files. This improves performance by allowing us to have less client components. It also improves development velocity by abstracting unimportant details and replacing them with a descriptive names. Props should be passed in as needed to expose information.

I didn't think about calling the API call directly, that might be the way forward. I was working in Region Select because that seemed to be the component that handled populating the region options.

File: components/ui/location-combobox.tsx
Comment (Luceium) [reply components/ui/location-combobox.tsx:1]: Thanks, I'll use the `DEFAULT_QUERY` to initialize things. The current implementation is already using `handleFindRegion`. I'm the one that added params to `handleFindRegion` and I made them optional, where the defaults used in the prev implementation will be used if no argument is provided. I didn't look to much into it yet but, I got an error when trying to use it. You can test this by:
1. Going to `localhost:3000`
2. Click on "select location"
image
3. Type in a non-cached value. (ex. "Egypt")
4. Click on "No location found. Click to search."
image

Also some additional clarification on the goal of the PR:
I want the combo box to have an option to search in case the region has not been previously cached. Then it should search as we do already. This combobox is meant to replace "Region Search Query"
image

File: components/search/RegionSearch/RegionSelect.tsx (hunk 1 above)
Comment (ray-sjsu) [inline components/search/RegionSearch/RegionSelect.tsx:86]: `LocationCombobox` is missing a crucial prop. It has no way of retrieve the API's result from `RegionSelect`. Add a new prop called `regionsArray`. This will pass the API data onto `LocationCombobox`.

```suggestion
      <LocationCombobox searchLocations={handleFindRegion} setRegionId={onRegionSelect} regionsArray={regionSearchOutput} />
```

File: components/ui/location-combobox.tsx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+type Location = {
+  name: string;
+  type: string;
+  regionId: string;
+  coordinates: [number, number];
+  country: string;
+  domain: string;
+};
+
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:28]:
```suggestion
import { APIRegionArrayFormatted } from "@/app/api/hotels/region/route";
```

`APIRegionArrayFormatted`. This is the region API's structure. Use this exact structure when creating cacheLocations. Or retrieve them from the API directly through Postman.

File: components/ui/location-combobox.tsx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  {
+    name: "San Jose State University, San Jose, California, United States",
+    type: "POI",
+    regionId: "6069527",
+    coordinates: [37.32890082806852, -121.8749250818403],
+    country: "United States",
+    domain: "US",
+  },
+];
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:120]:
```suggestion
const cachedLocations: APIRegionArrayFormatted = [
  {
    region_id: "602703",
    type: "CITY",
    regionNames: {
      fullName: "San Jose, California, United States",
      shortName: "San Jose",
      displayName: "San Jose, California, United States",
      primaryDisplayName: "San Jose",
      secondaryDisplayName: "California, United States",
      lastSearchName: "San Jose, California, United States",
    },
    coordinates: {
      lat: "37.3354",
      long: "-121.891907",
    },
    country: {
      name: "United States",
      domain: "US",
    },
  },
  {
    region_id: "4409939",
    type: "AIRPORT",
    regionNames: {
      fullName:
        "San Jose (SJC - Norman Y. Mineta San Jose Intl.), California, United States",
      shortName: "SJC",
      displayName: "San Jose Airport",
      primaryDisplayName: "Norman Y. Mineta San Jose Intl.",
      secondaryDisplayName: "San Jose, California, United States",
      lastSearchName: "San Jose (SJC - Norman Y. Mineta San Jose Intl.)",
    },
    coordinates: {
      lat: "37.369739",
      long: "-121.929225",
    },
    country: {
      name: "United States",
      domain: "US",
    },
  },
  {
    region_id: "3177",
    type: "CITY",
    regionNames: {
      fullName: "San José, San José Province, Costa Rica",
      shortName: "San José",
      displayName: "San José, Costa Rica",
      primaryDisplayName: "San José",
      secondaryDisplayName: "San José Province, Costa Rica",
      lastSearchName: "San José, Costa Rica",
    },
    coordinates: {
      lat: "9.93286",
      long: "-84.079559",
    },
    country: {
      name: "Costa Rica",
      domain: "CR",
    },
  }
];
```

`cacheLocations` now in proper API format.

File: components/ui/location-combobox.tsx [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+type LocationComboboxProps = {
+  searchLocations: (query: string) => void;
+  setRegionId: (regionId: string) => void;
+};
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:125]:
```suggestion
type LocationComboboxProps = {
  searchLocations: (query: string) => void;
  setRegionId: (regionId: string) => void;
  regionsArray: APIRegionArrayFormatted | null;
};
```

Add new prop `regionsArray` to pass API data into `LocationCombobox`

File: components/ui/location-combobox.tsx [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+      <PopoverTrigger asChild>
+        <Button
+          variant="outline"
+          role="combobox"
+          aria-expanded={open}
+          className="w-[200px] justify-between"
+        >
+          {value
+            ? cachedLocations.find((loc) => loc.name === value)?.name
+            : "Select Location..."}
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:146]: Modify this to fit structure `APIRegionArrayFormatted`.

File: components/ui/location-combobox.tsx [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        </Button>
+      </PopoverTrigger>
+      <PopoverContent className="w-[200px] p-0">
+        <Command>
+          <CommandInput placeholder="Search locations..." />
+          <CommandList>
+            <CommandEmpty>
+              <button onClick={handleSearch}>
+                No location found. Click to search.
+              </button>
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:157]:
```suggestion
              <button onClick={() => handleSearch(value)}>
                No location found. Click to search.
              </button>
```

File: components/ui/location-combobox.tsx [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  setRegionId: (regionId: string) => void;
+};
+
+export default function LocationCombobox(props: LocationComboboxProps) {
+  const [open, setOpen] = React.useState(false);
+  const [value, setValue] = React.useState("");
+  const handleSearch = () => {
+    setValue("");
+    props.searchLocations(value);
+  };
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:133]:
```suggestion
  const handleSearch = (value: string) => {
      setValue(value);
      props.searchLocations(value);
  };
```

The input for the API `searchLocations` is not being updated properly.

File: components/ui/location-combobox.tsx [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        >
+          {value
+            ? cachedLocations.find((loc) => loc.name === value)?.name
+            : "Select Location..."}
+          <ChevronsUpDown className="opacity-50" />
+        </Button>
+      </PopoverTrigger>
+      <PopoverContent className="w-[200px] p-0">
+        <Command>
+          <CommandInput placeholder="Search locations..." />
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:152]: Modify `CommandInput` (the searchbar), so it actually passes the user input's to `handleChange`.

File: components/ui/location-combobox.tsx (hunk 7 above)
Comment (ray-sjsu) [reply components/ui/location-combobox.tsx:133]: Also, add another useState hook to store the region id that the user selected from the API or cached locations.

**Review by ray-sjsu — CHANGES_REQUESTED** [review]
The region search API outputs a specific structure: `APIRegionArrayFormatted`. Use this in `location-combobox.tsx` when creating cached regions and for inputs/outputs.

`location-combobox.tsx` is missing a prop to recieve API data and not updating its input props `searchLocations` to pass user data to the API. Also, update `setRegionId` prop.

**Comment by Luceium** [comment]
@KingSand08 I made it work with your navbar. It's a bit scuffed right now but it works.

**Comment by Luceium** [comment]
Loading indicators are the next step for this component.

File: components/ui/location-combobox.tsx [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+      className={`relative ${
+        open ? "rounded-lg rounded-b-none" : "rounded-lg"
+      } text-neutral dark:text-gray-100 dark:bg-gray-800 border-2 border-primary focus:outline-none focus:ring-2 focus:ring-blue-500`}
+    >
+      <CommandInput
+        placeholder="Search locations..."
+        onValueChange={setQuery}
+        onKeyDown={(e) => {
+          if (e.key == "Enter") findRegion();
+          else if (e.key == "Escape") setOpen(false);
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:109]:
```suggestion
          if (e.key === "Enter") findRegion();
          else if (e.key === "Escape") setOpen(false);
```

Always use `===` or `!==`. Blame javascript for this.

File: components/ui/location-combobox.tsx [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+  DEFAULT_DOMAIN,
+  DEFAULT_LOCALE,
+} from "@/lib/rapid-hotel-api/constants/USER_OPTIONS";
+import { JSONToURLSearchParams } from "@/lib/rapid-hotel-api/APIFunctions";
+import { REGION_SEARCH_API_URL } from "@/lib/rapid-hotel-api/constants/ROUTES";
+import { APIRegionArrayFormatted } from "@/app/api/hotels/region/route";
+import { useState, useContext } from "react";
+import { RegionContext } from "../providers/RegionProvider";
+import { CommandEmpty } from "cmdk";
+
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:23]:
```suggestion
const dropdownRef = React.useRef<HTMLDivElement>(null);

```

File: components/ui/location-combobox.tsx [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+  },
+];
+
+export default function LocationCombobox() {
+  const [value, setValue] = useContext(RegionContext);
+  const [open, setOpen] = useState(false);
+  const [ops, setOps] = useState(cachedLocations);
+  const [query, setQuery] = useState("");
+  const findRegion = () => handleFindRegion(query, ops, setOps);
+
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:95]:
```suggestion
  const handleBlur = (e: React.FocusEvent) => {
    // Check if focus is still within the dropdown
    if (!dropdownRef.current?.contains(e.relatedTarget as Node)) {
      setOpen(false);
    }
  };
  
  
```

Should fix clicking twice on dropdown. Part 2.

File: components/ui/location-combobox.tsx [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+          <CommandEmpty className="py-3 text-center text-sm w-full">
+            Press Enter to search for more locations.
+          </CommandEmpty>
+          <CommandGroup>
+            {ops.map((loc) => (
+              <CommandItem
+                key={loc.region_id}
+                value={loc.regionNames.displayName}
+                onSelect={(currentValue) => {
+                  const isDeselect = currentValue == value;
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:124]:
```suggestion
                  const isDeselect = currentValue === value?.name;
```

Comment (ray-sjsu) [reply components/ui/location-combobox.tsx:124]: `currentValue` is of type string.
`value` is of type RegionContextType (an object {})

File: components/ui/location-combobox.tsx [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+                    name: loc.regionNames.displayName,
+                  });
+                  setOpen(false);
+                }}
+              >
+                {loc.regionNames.displayName}
+                <Check
+                  className={cn(
+                    "ml-auto",
+                    value === loc.regionNames.displayName
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:140]:
```suggestion
                    value?.name === loc.regionNames.displayName
```

`loc.regionNames.displayName` is of type string.
`value` is of type RegionContextType (an object {region_id: string, name: string})

File: components/ui/location-combobox.tsx [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+  const [value, setValue] = useContext(RegionContext);
+  const [open, setOpen] = useState(false);
+  const [ops, setOps] = useState(cachedLocations);
+  const [query, setQuery] = useState("");
+  const findRegion = () => handleFindRegion(query, ops, setOps);
+
+  return (
+    <Command
+      onFocus={() => setOpen(true)}
+      onBlur={() => setOpen(false)}
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:99]:
```suggestion
    <Command
      href={dropdownRef}
      onFocus={() => setOpen(true)}
      onBlur={(handleBlur)}
```

Should fix clicking twice on dropdown. Part 3.

**Review by ray-sjsu — CHANGES_REQUESTED** [review]
Found a weird bug.
1. Search for San Jose (or any location) in the searchbox.
2. Click on one of the regions. Does not close the dropbox and select a region (update the title).
You have to click twice on the dropdown item in order for the page to actually get a region.

I recommend a fix for this particular issue ("clicking twice on dropdown item"). There are some possible type errors.

Some recommendations:
1. Selecting a region should autofill the search box. Then, clicking on the search box should clear the text, assuming you're searching for a new region.
2. Within the dropdown, clicking on an already selected region unchecks the item and clears the selected region. Is this intended behavior or is it supposed to close the dropdown and keep the current region since the region is already selected.

**Review by ray-sjsu — APPROVED** [review]
Currently, the implementation of combobox replaces the old `RegionSelect.tsx` component.

Putting it in the Navbar and having it available for every page doesn't make sense at the moment. `LocationCombobox` does not persist data between reloads (like navigating to different pages) At it's current implementation, its only good for the browsing/search page

I moved `LocationCombobox` from navbar to the browsing/search page (where `RegionSelect` used to be). Also, fixed a dropdown bug + add default values to region context.

File: components/providers/RegionProvider.tsx [hunk 15]
```diff
@@ -0,0 +1,24 @@
+"use client";
+import { DEFAULT_SELECTED_REGION } from "@/lib/rapid-hotel-api/constants/USER_OPTIONS";
+import { createContext, ReactNode, useState } from "react";
```
Comment (ray-sjsu) [inline components/providers/RegionProvider.tsx:3]:
```suggestion
import { createContext, ReactNode, useEffect, useState } from "react";
```

File: components/providers/RegionProvider.tsx [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+
+export const RegionContext = createContext<
+  [RegionContextType, React.Dispatch<React.SetStateAction<RegionContextType>>]
+>([{}, () => {}]);
+
+const RegionProvider = ({ children }: { children: ReactNode }) => {
+  const [region, setRegion] = useState<RegionContextType>({
+    region_id: DEFAULT_SELECTED_REGION.region_id,
+    name: DEFAULT_SELECTED_REGION.regionNames.displayName
+  });
```
[the remark this answers is not in the document]
Comment (ray-sjsu) [reply components/providers/RegionProvider.tsx:15]: https://www.hotels.com/ This is similar behavior to hotels.com (at least in their browsing page).

File: components/ui/location-combobox.tsx [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+      } text-neutral dark:text-gray-100 dark:bg-gray-800 border-2 border-primary focus:outline-none focus:ring-2 focus:ring-blue-500`}
+    >
+      <CommandInput
+        placeholder="Search locations..."
+        onValueChange={setQuery}
+        onKeyDown={(e) => {
+          if (e.key === "Enter") findRegion();
+          else if (e.key === "Escape") setOpen(false);
+        }}
+      />
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:122]: I think this is the input/search box.

Clicking on a selected location should autofill the searchbox with the selected region's name. I believe query in `Line 144` should have the fully displayname of the region.

File: components/ui/location-combobox.tsx [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+      domain: "CR",
+    },
+  },
+];
+
+export default function LocationCombobox() {
+  const [value, setValue] = useContext(RegionContext);
+  const [open, setOpen] = useState(false);
+  const [ops, setOps] = useState(cachedLocations);
+  const [query, setQuery] = useState("");
```
Comment (ray-sjsu) [inline components/ui/location-combobox.tsx:93]:
```suggestion
  const [query, setQuery] = useState(value?.name || "");
```

File: components/providers/RegionProvider.tsx (hunk 16 above)
Comment (ray-sjsu) [inline components/providers/RegionProvider.tsx:15]:
```suggestion
  const [region, setRegion] = useState<RegionContextType>(() => {
    const storedRegion = localStorage.getItem("region");
    return storedRegion
      ? JSON.parse(storedRegion)
      : {
          region_id: DEFAULT_SELECTED_REGION.region_id,
          name: DEFAULT_SELECTED_REGION.regionNames.displayName,
        };
  });

  useEffect(() => {
    localStorage.setItem("region", JSON.stringify(region));
  }, [region]);
```

Store region data in `localStorage`, so its actually persistent.
image

**Review by ray-sjsu — CHANGES_REQUESTED** [review]
Add these changes, so `RegionContext` is actually persistent between page reloads and navigation.

Then, it should be good to go.

**Comment by Luceium** [comment]
I looked over it and we persist it for navigation between pages but you're right refresh does remove it. I'm not sure if this is a priority or even intended feature. Instead of local storage it would be better in the url params. But I think we should save that for after MMFs @ray-sjsu .

**Review by ray-sjsu — APPROVED** [review]
